### ISSUE TITLE
Netherquark's bug reports/feature requests megathread

### ISSUE THREAD

**Issue by Netherquark** [icidasset/diffuse#272]
I am creating a GitHub issue with all the bug reports & feature requests instead of sending them to @icidasset 's Twitter so it's more easily trackable and he won't need to check social media while working lol
I'll edit this issue and add other things as I find them or think of them

_All of these are reported on latest Chrome on Windows on nightly.diffuse.sh with Dropbox as the user layer and GDrive as the music layer. Chrome on Android is the browser in question if specifically mentioned._

**Feature requests:**
- [ ] Mega support
- [ ] Cached song management is very basic, could be greatly improved
- [ ] Serverless implementation of Spotify Connect
- [ ] User layer only syncs on Diffuse startup, doesn't support live sync
- [ ] The attempts to fix the offline bug above slowed down Diffuse load times, which were already not as fast as a local app (IDK what might fix it)
- [ ] When caching tracks, Diffuse caches one song at a time, greatly lengthening the time required for it to finish caching everything. An implementation of the "download as a zip" function would help speed this up
- [ ] Artist view and album view should function like playlists, setting the artist/album in view as the pool to play the next song from as is the case with playlists

**Bugs:**
- [ ] Diffuse gets stuck on Loading Application if the encryption key is wrong.
- [ ] Diffuse doesn't [scale ](https://photos.app.goo.gl/X34FuWEnsyEBsNY96) well on 2400x1080p display on Chrome on Android [you can see the bottom area empty in the screenshot linked]
- [ ] Diffuse doesn't scale to a display of less than 1600x1080p when using the split-screen feature on Chrome on Android. The UI is scrollable if that is the case.
- [ ] When right-clicking/long-pressing on a UI element such as a song or a playlist to select an option such as download as a zip file, the [element behind it is also clicked](https://photos.app.goo.gl/hdQyHJiW2cETgHga9). [For example, in the screenshot linked, if I click download as a zip, it'll open Sleep or a struggle for bliss along with starting the download]
- [ ] On Chrome on Android, the cached tracks don't show up under the app as storage, they show up under the Chrome app[[screenshots](https://photos.app.goo.gl/qqvrHoyR13Zq4Hn57)]. However, this isn't the case on Chrome on Windows
- [ ] The Grammarly extension breaks the "Enter encryption key" page
- [ ] Diffuse shows "No Tracks" even if tracks are cached when offline while opening the app
- [ ] Sometimes the tracks show up as processed but the names and metadata is gibberish (a quick fix is alt right-clicking, but that is per track and doesn't work for all tracks at once)
- [ ] Occasionally Diffuse gives a 502 Bad Gateway due to miscellaneous hosting issues and because of the offline no tracks bug, this renders the app unusable till the hosting issues are fixed
- [ ] Sometimes when using shuffle, a song repeats for seemingly forever before a different song is played (the queue is filled with multiple repetitions of the same song randomly)[[Screen record](https://photos.app.goo.gl/U7T4WMBjpV4fcTyR7)]
- [ ] When a song from a playlist is directly played by double-clicking on it, Diffuse glitches out [[Screen record](https://photos.app.goo.gl/355VH9sTFeog4BnJ7)]
- [ ] The last.fm integration shows songs being played but doesn't register them in the last.fm history unless they're played from start to end, without any skipping, etc.
- [ ] If a batch cache is started (eg cache playlist, cache selected tracks) then if the cache isn't properly cached, diffuse shows "can't play this track because the browser doesn't support it" until the song is manually removed from cache and readded to cache
- [ ] [Question] Will Diffuse still work on Chrome once the third party cookie support is removed? Cause I've noticed that user layer login credentials are stored in cookies [not sure though]
- [ ] UI unreadable in some cases [[Screenshot](https://photos.app.goo.gl/JMwd1wuTsSTdKqEP6)]
- [ ] UI design isn't very logical in some cases [Eg. you cannot manage a playlist by caching it or renaming it etc while you're listening to it because of the ✅]

**Comment by icidasset** [comment]
Thanks for all the feedback! ☺️ 
I couldn't keep up anymore with the Twitter messages, so thanks for making a summary 🙏 

Few things:
1. The last.fm scrobble only comes through if the track is played for 50%, unless your track is only 2 min long, or something like that. This is required by last.fm
2. > [Question] Will Diffuse still work on Chrome once the third party cookie support is removed? Cause I've noticed that user layer login credentials are stored in cookies [not sure though]

    I'm not using cookies in any way, so it should work just fine. What gave you the impression cookies are used? 🤔

**Comment by Netherquark** [comment]
> Thanks for all the feedback! ☺️ 

My pleasure!

> Few things:
> 1. The last.fm scrobble only comes through if the track is played for 50%, unless your track is only 2 min long, or something like that. This is required by last.fm

Ah okay

> 2. > [Question] Will Diffuse still work on Chrome once the third party cookie support is removed? Cause I've noticed that user layer login credentials are stored in cookies [not sure though]
> 
>     I'm not using cookies in any way, so it should work just fine. What gave you the impression cookies are used? 🤔 

Clearing cookies & clearing storage is what I do usually if I want to reset diffuse (eg if I type in the wrong credentials). Also diffuse uses Google and dropbox cookies during login? It's always a good thing if diffuse doesn't need cookies tho

**Comment by icidasset** [comment]
> If I have a song file, say "Alien Ant Farm - Smooth Criminal.mp3", and I delete that file from the drive, and replace it with "Alien Ant Farm - Smooth Criminal.mp3", but at a higher bitrate, as a part of upgrading my library bitrate. What happens is, diffuse directs the old processed entry to this new file, while also processing the new file, resulting in duplicate entries. The only possible way to fix this is to export diffuse.json, edit it to remove all mentions of the duplicate file in question, import it into diffuse, and process the sources, where it just creates one valid entry.

That's quite a special use case 🤔  I can't really think of a great solution. But for now you can do one of the following:
- Delete the files on your source/drive, click process, wait until finished, upload new files, process again.
- Upload new files, alt + right click on each changed track, and choose "sync tags"

The tags (audio metadata) are different but same filename, right?

**Comment by icidasset** [comment]
> After a certain point of time, Diffuse gives the "Can't play this track" error on every song that isn't cached till it is reopened. [Screenshot attached.] This happens even on a 250mb/s LAN connection, so it can't be related to internet issues. The same thing occurs on a very low-speed internet connection (1.5mbit/s)

Huh interesting, I guess there's a rate limit on Google Drive. I don't think there's much I can do about that, seeing that it's a limitation of Google Drive. I could show a message to the user however.

**Comment by Netherquark** [comment]
> That's quite a special use case 🤔 I can't really think of a great solution. But for now, you can do one of the following:
> 
> * Delete the files on your source/drive, click process, wait until finished, upload new files, process again.
> * Upload new files, alt + right-click on each changed track, and choose "sync tags"
> 
> The tags (audio metadata) are different but the same filename, right?

yep same file name different metadata (eg the higher-res version has album art, album metadata, genre, etc)
i think the best solution to this would be to give the user a scan button, which would automatically look through for inconsistencies (eg multiple listings as per the Smooth Criminal example, tracks that show up as cached but aren't actually cached)
because the best way I found of fixing the above issue on a per-song basis was to ctrl f the diffuse JSON with the song name, and remove all entries, so it was reprocessed.

**Comment by Netherquark** [comment]
> > After a certain point of time, Diffuse gives the "Can't play this track" error on every song that isn't cached till it is reopened. [Screenshot attached.] This happens even on a 250mb/s LAN connection, so it can't be related to internet issues. The same thing occurs on a very low-speed internet connection (1.5mbit/s)
> 
> Huh interesting, I guess there's a rate limit on Google Drive. I don't think there's much I can do about that, seeing that it's a limitation of Google Drive. I could show a message to the user however.

maybe that'll work

**Comment by Glandos** [comment]
Subsonic API is already consumed in [Jamstash](https://github.com/tsquillario/Jamstash), but I don't know if the logic can be extracted easily.
